Change default SQL mapping for `VARCAHR` from `Utf8` to  `Utf8View`

zhuqi-lucas opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes [#15096](https://github.com/apache/datafusion/issues/15096)

## Rationale for this change

Setting Varchar default mapping to utf8view, and change testing result.
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?
Setting Varchar default mapping to utf8view, and change testing result.
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

Yes, new testing/slt testing result generated.
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

No, the internal datatype will change to utf8view.

zhuqi-lucas commented:
Interesting, sqllogic testing successfully for my local mac but CI failed...

zhuqi-lucas commented:
Resolved all problems now including the slt testing.

I think the PR is ready for benchmark progress to see if switching to Utf8View by default slows some queries down.  cc @alamb Thanks a lot.

alamb commented:
🤖 `./gh_compare_branch.sh` [Benchmark Script](https://github.com/alamb/datafusion-benchmarking/blob/main/gh_compare_branch.sh) Running
Linux aal-dev 6.11.0-1013-gcp #13~24.04.1-Ubuntu SMP Wed Apr  2 16:34:16 UTC 2025 x86_64 x86_64 x86_64 GNU/Linux
Comparing varchar_default_ut8view (5d9c537c446073c433bb4e3299bf9aa775fa0b60) to f3aed4a9bf988f5c0fdc3be845d57c0fe253cd45 [diff](https://github.com/apache/datafusion/compare/f3aed4a9bf988f5c0fdc3be845d57c0fe253cd45..5d9c537c446073c433bb4e3299bf9aa775fa0b60)
Benchmarks: tpch_mem clickbench_partitioned clickbench_extended
Results will be posted here when complete

alamb commented:
> Resolved all problems now including the slt testing.
> 
> I think the PR is ready for benchmark progress to see if switching to Utf8View by default slows some queries down. cc @alamb Thanks a lot.

THanks @zhuqi-lucas  -- I have started it up

alamb commented:
🤖: Benchmark completed

<details><summary>Details</summary>
<p>


```
Comparing HEAD and varchar_default_ut8view
--------------------
Benchmark clickbench_extended.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃       HEAD ┃ varchar_default_ut8view ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 0     │  1898.26ms │               1810.38ms │     no change │
│ QQuery 1     │   658.97ms │                670.82ms │     no change │
│ QQuery 2     │  1406.85ms │               1396.48ms │     no change │
│ QQuery 3     │   709.94ms │                713.90ms │     no change │
│ QQuery 4     │  1429.02ms │               1398.58ms │     no change │
│ QQuery 5     │ 15171.65ms │              15018.76ms │     no change │
│ QQuery 6     │  2057.97ms │               1992.46ms │     no change │
│ QQuery 7     │  2336.53ms │               2061.27ms │ +1.13x faster │
│ QQuery 8     │   839.18ms │                837.10ms │     no change │
└──────────────┴────────────┴─────────────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary                      ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (HEAD)                      │ 26508.38ms │
│ Total Time (varchar_default_ut8view)   │ 25899.76ms │
│ Average Time (HEAD)                    │  2945.38ms │
│ Average Time (varchar_default_ut8view) │  2877.75ms │
│ Queries Faster                         │          1 │
│ Queries Slower                         │          0 │
│ Queries with No Change                 │          8 │
└────────────────────────────────────────┴────────────┘
--------------------
Benchmark clickbench_partitioned.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃       HEAD ┃ varchar_default_ut8view ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 0     │    15.20ms │                 15.35ms │     no change │
│ QQuery 1     │    33.35ms │                 33.29ms │     no change │
│ QQuery 2     │    83.06ms │                 79.39ms │     no change │
│ QQuery 3     │    96.03ms │                 95.15ms │     no change │
│ QQuery 4     │   603.39ms │                590.48ms │     no change │
│ QQuery 5     │   851.82ms │                855.46ms │     no change │
│ QQuery 6     │    23.71ms │                 22.55ms │     no change │
│ QQuery 7     │    36.83ms │                 37.07ms │     no change │
│ QQuery 8     │   933.66ms │                916.79ms │     no change │
│ QQuery 9     │  1192.79ms │               1193.28ms │     no change │
│ QQuery 10    │   260.86ms │                262.38ms │     no change │
│ QQuery 11    │   299.90ms │                293.48ms │     no change │
│ QQuery 12    │   906.00ms │                898.14ms │     no change │
│ QQuery 13    │  1344.05ms │               1220.93ms │ +1.10x faster │
│ QQuery 14    │   842.98ms │                823.52ms │     no change │
│ QQuery 15    │   842.44ms │                835.95ms │     no change │
│ QQuery 16    │  1725.19ms │               1718.98ms │     no change │
│ QQuery 17    │  1598.00ms │               1614.71ms │     no change │
│ QQuery 18    │  3092.64ms │               3093.01ms │     no change │
│ QQuery 19    │    81.09ms │                 81.79ms │     no change │
│ QQuery 20    │  1111.88ms │               1115.69ms │     no change │
│ QQuery 21    │  1298.42ms │               1296.34ms │     no change │
│ QQuery 22    │  2165.31ms │               2174.91ms │     no change │
│ QQuery 23    │  8034.33ms │               7921.79ms │     no change │
│ QQuery 24    │   458.46ms │                457.26ms │     no change │
│ QQuery 25    │   391.76ms │                372.64ms │     no change │
│ QQuery 26    │   528.56ms │                519.61ms │     no change │
│ QQuery 27    │  1595.08ms │               1586.78ms │     no change │
│ QQuery 28    │ 12566.00ms │              12405.74ms │     no change │
│ QQuery 29    │   535.24ms │                532.68ms │     no change │
│ QQuery 30    │   796.06ms │                807.39ms │     no change │
│ QQuery 31    │   863.55ms │                857.96ms │     no change │
│ QQuery 32    │  2633.13ms │               2621.41ms │     no change │
│ QQuery 33    │  3300.20ms │               3347.80ms │     no change │
│ QQuery 34    │  3429.42ms │               3335.38ms │     no change │
│ QQuery 35    │  1338.57ms │               1340.32ms │     no change │
│ QQuery 36    │   123.48ms │                118.34ms │     no change │
│ QQuery 37    │    55.36ms │                 56.67ms │     no change │
│ QQuery 38    │   119.30ms │                117.49ms │     no change │
│ QQuery 39    │   191.24ms │                187.82ms │     no change │
│ QQuery 40    │    46.61ms │                 47.57ms │     no change │
│ QQuery 41    │    45.34ms │                 44.25ms │     no change │
│ QQuery 42    │    39.10ms │                 38.06ms │     no change │
└──────────────┴────────────┴─────────────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary                      ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (HEAD)                      │ 56529.39ms │
│ Total Time (varchar_default_ut8view)   │ 55985.60ms │
│ Average Time (HEAD)                    │  1314.64ms │
│ Average Time (varchar_default_ut8view) │  1301.99ms │
│ Queries Faster                         │          1 │
│ Queries Slower                         │          0 │
│ Queries with No Change                 │         42 │
└────────────────────────────────────────┴────────────┘
--------------------
Benchmark tpch_mem_sf1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Query        ┃     HEAD ┃ varchar_default_ut8view ┃    Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ QQuery 1     │ 122.52ms │                121.90ms │ no change │
│ QQuery 2     │  22.09ms │                 22.25ms │ no change │
│ QQuery 3     │  33.67ms │                 34.12ms │ no change │
│ QQuery 4     │  19.78ms │                 19.82ms │ no change │
│ QQuery 5     │  50.90ms │                 52.52ms │ no change │
│ QQuery 6     │  11.94ms │                 11.89ms │ no change │
│ QQuery 7     │  94.09ms │                 93.62ms │ no change │
│ QQuery 8     │  25.99ms │                 25.51ms │ no change │
│ QQuery 9     │  60.34ms │                 59.78ms │ no change │
│ QQuery 10    │  56.02ms │                 56.36ms │ no change │
│ QQuery 11    │  11.44ms │                 11.54ms │ no change │
│ QQuery 12    │  41.27ms │                 41.51ms │ no change │
│ QQuery 13    │  27.81ms │                 27.49ms │ no change │
│ QQuery 14    │   9.53ms │                  9.54ms │ no change │
│ QQuery 15    │  23.84ms │                 24.62ms │ no change │
│ QQuery 16    │  21.35ms │                 22.00ms │ no change │
│ QQuery 17    │  95.29ms │                 95.13ms │ no change │
│ QQuery 18    │ 207.37ms │                212.10ms │ no change │
│ QQuery 19    │  26.88ms │                 26.29ms │ no change │
│ QQuery 20    │  38.11ms │                 37.17ms │ no change │
│ QQuery 21    │ 164.65ms │                161.44ms │ no change │
│ QQuery 22    │  15.92ms │                 16.45ms │ no change │
└──────────────┴──────────┴─────────────────────────┴───────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Benchmark Summary                      ┃           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Total Time (HEAD)                      │ 1180.80ms │
│ Total Time (varchar_default_ut8view)   │ 1183.07ms │
│ Average Time (HEAD)                    │   53.67ms │
│ Average Time (varchar_default_ut8view) │   53.78ms │
│ Queries Faster                         │         0 │
│ Queries Slower                         │         0 │
│ Queries with No Change                 │        22 │
└────────────────────────────────────────┴───────────┘
```


</p>
</details>

zhuqi-lucas commented:
Thank you @alamb , it seems no regression from the benchmark result!

alamb commented on datafusion/core/tests/user_defined/user_defined_plan.rs:
```diff
@@ -796,22 +798,30 @@ fn accumulate_batch(
+
     // Assuming the input columns are
-    // column[0]: customer_id / UTF8
+    // column[0]: customer_id / UTF8 or UTF8View
```
I feel like the example would be simpler if it just used `Utf8View`. Could we do that and avoid adding the logic to handle both types?

zhuqi-lucas replied:
Good suggestion! Thank you @alamb , addressed in latest PR.

zhuqi-lucas commented:
> Thank you @zhuqi-lucas -- this is great
> 
> FYI @jayzhan211 , @timsaucer @xudong963 and @comphead
> 
> I have a few suggestions on how to simplify the examples which would be nice to get in prior to merge but I don't think it is needed

Thank you @alamb for review, i also addressed the suggestions about example in latest PR.

zhuqi-lucas commented:
Thank you @xudong963 for review!

xudong963 commented:
Let's go, thank you all

alamb commented:
Thanks again @zhuqi-lucas and @xudong963 

I have also made a PR to add a mention of this change to the upgrade guide:
- https://github.com/apache/datafusion/pull/16216
